src/object/cli_obj.c

```diff
@@ -2272,16 +2273,24 @@ obj_comp_cb(tse_task_t *task, void *data)
 		 * shard, such as -DER_TIMEDOUT or daos_crt_network_error().
 		 */
 
-		if (!obj_auxi->spec_shard ||
+		if ((!obj_auxi->spec_shard && !obj_auxi->no_retry) ||
 		    task->dt_result != -DER_INPROGRESS)
 			obj_auxi->io_retry = 1;
```

Comment:
you want to use the no_retry flag to force it don't retry right? but above if (... || task->dt_result != -DER_INPROGRESS) will be true as the dt_result is DER_CSUM

actually looks like you need not change the code here, because you handled the DER_CSUM case below.